### DISCUSSION TITLE
Strange queries from dnscrypt-proxy.exe.

### DISCUSSION THREAD

**Discussion by maestropss** [DNSCrypt/dnscrypt-proxy#2201] [Q&A] [answered]
Hello friends.

Every night at 4:18 there are four queries from dnscrypt-proxy.exe to 53 port of 9.9.9.9 cloudflare server. Two of them use TCP and the other two of them use UDP protocols. The strange thing is that:

1. 9.9.9.9 is not written anywhere in the configuration file (dnscrypt-proxy.toml)

2. I do not use 9.9.9.9 as a system DNS resolver. There is only one network interface and it uses 127.0.0.1 for IPv4 . IPv6 is disabled in the system.

3.  bootstrap_resolvers and netprobe_address parameteres in dnscrypt-proxy.toml are commented (disabled). Resolvers, relays, certificates are updated from cache or through dnscrypt resolvers after they become available. ISP DNS hijacking, you know...

4. cert_refresh_delay is set to 60 minutes, log file shows no warnings or errors. So it should not be related to these parameters.

5. refresh_delay of resolvers and relays are set to 72 hours (not 24 hours) so it should not be related to these parameters.

6. The log file contains the following (it happens every 24 hours at 4:18 at night). 

[2022-09-18 04:18:21] [NOTICE] System DNS configuration not usable yet, exceptionally resolving [download.dnscrypt.info] using bootstrap resolvers over tcp
[2022-09-18 04:18:21] [NOTICE] Bootstrap resolvers didn't respond - Trying with the system resolver as a last resort
[2022-09-18 04:18:26] [NOTICE] System DNS configuration not usable yet, exceptionally resolving [raw.githubusercontent.com] using bootstrap resolvers over tcp
[2022-09-18 04:18:26] [NOTICE] Bootstrap resolvers didn't respond - Trying with the system resolver as a last resort

7. Windows DNScache service is disabled. UDP, TCP 53 port is not allowed for dnscrypt-proxy.exe in my configuration. It is not blocked though, so if dnscrypt proxy needs query DNS at this port it should pass through the system DNS, which is 127.0.0.1. This is how it works for all other software. 

The questions:

A. Why does dnscrypt-proxy.exe query DNS resolver at 53 port every 24 hours? If it tries to update relays and resolvers list, why does it NOT use available dnscrypt resolvers? And why does it happen every 24 hours while refresh_delay for resolvers and relays are set to 72 hours?

B. Why does dnscrypt-proxy.exe try to resolve something by using 9.9.9.9 resolver while this resolver is not mentioned anywhere in the system and dncrypt-proxy configuration?

Thanks in advance

**Comment by jedisct1** [comment]
If you don't have a usable system DNS, some DNS resolver is still required to resolve the names of the list sources, as well as resolve DoH server names. So, bootstrap resolvers are used in that case. If you don't configure any, the defaults are used, which are `9.9.9.9` (Quad9, not Cloudflare).

**Reply by maestropss** [reply]
Thank you. 
<If you don't have a usable system DNS>
But i do. DNS crypt servers are available. According to dnscrypt-proxy.toml comments, "They (Bootstrap resolvers，non-encrypted DNS resolvers) will not be used after the proxy already has at least one usable secure resolver".

In addition, it is not quite clear why this query is made every 24 hours. Is it hard coded?

Thank you

**Comment by ghost** [comment]
This looks like the typical forced reconnect some ISPs do once a day.

**Comment by lifenjoiner** [comment, marked as answer]
I think it is over concerned:
1. The hosts resolved through the bootstrap resolvers are those in the config file: source domains and DoH hosts without addresses. But none of those queries `dnscrypt-proxy` received are forwarded to them at all.
2. Those hosts are looked up at the startup and when the cached IPs are expired.
3. Bootstrap resolvers are necessary, for the situation when a user uses only DoH and all of they have no address preset.
4. In fact, most of the time, the system DNS leads to dnscrypt-proxy itself, if you set up dnscrypt-proxy in the system DNS chain. And it is the bootstrap resolver too.

When you ignored system DNS and removed all bootstrap_resolvers, it could cause problem to the situation in item 3. So, embedded default bootstrap resolvers are necessary.

I totally understand your concern about querying to those non-encrypted DNS. https://github.com/DNSCrypt/dnscrypt-proxy/pull/2204 is try to do some improvements.

**Reply by llightcb** [reply]
hello.

i still don't understand what exactly is happening here (and why):

yesterday:

Jul 26 14:00:03 System DNS configuration not usable yet, exceptionally resolving [download.dnscrypt.net] using bootstrap resolvers over tcp
Jul 26 14:00:07 Bootstrap resolvers didn't respond - Trying with the system resolver as a last resort
Jul 26 14:00:07 System DNS configuration not usable yet, exceptionally resolving [raw.githubusercontent.com] using bootstrap resolvers over tcp
Jul 26 14:00:11 Bootstrap resolvers didn't respond - Trying with the system resolver as a last resort
Jul 26 14:00:11 System DNS configuration not usable yet, exceptionally resolving [download.dnscrypt.info] using bootstrap resolvers over tcp
Jul 26 14:00:15 Bootstrap resolvers didn't respond - Trying with the system resolver as a last resort
..
Jul 26 14:30:01 Server with the lowest initial latency: cloudflare-security (rtt: 31ms)

after 24h:

Jul 27 14:01:09 Bootstrap resolvers didn't respond - Trying with the system resolver as a last resort
Jul 27 14:01:09 System DNS configuration not usable yet, exceptionally resolving [download.dnscrypt.info] using bootstrap resolvers over tcp
Jul 27 14:01:13 Bootstrap resolvers didn't respond - Trying with the system resolver as a last resort
..
Jul 27 16:51:05 Server with the lowest initial latency: cloudflare-security (rtt: 31ms)

->
bootstrap resolvers are not commented, ignore_system_dns = true, more or less everything = default .toml.

**Reply by lifenjoiner** [reply]
I can't get any clue from your information.

More over, these are in one group:
```
Jul 26 14:00:07 Bootstrap resolvers didn't respond - Trying with the system resolver as a last resort
Jul 26 14:00:07 System DNS configuration not usable yet, exceptionally resolving [raw.githubusercontent.com] using bootstrap resolvers over tcp
...
```
Meaning, while updating sources, it tried to resolve the host,
1. using bootstrap resolvers first, but failed: `didn't respond`;
2. although you disabled system DNS, but there is no other ways left. So, it still tried it as a last resort, but still failed: `not usable yet`.
3. `...`, as time flies, it retried, the bootstrap resolvers worked at some time, and then the sources are fetched finally.

Does your bootstrap resolvers work well?
What are your bootstrap resolvers and system DNS?
`System DNS configuration not usable yet` should only happen at startup, does it happen every 24h on the fly?
What's your system environment?

That's the direction that you can troubleshoot.

**Reply by llightcb** [reply]
thanks for your friendly response.

1. bootstrap resolvers didn't respond
_"using bootstrap resolvers first, but failed: didn't respond"_
"the bootstrap resolvers worked -->_at-some-time_"
**q**: but **why**? 1st and 2nd. as aready mentioned above: more or less default _.toml_.

so: `bootstrap_resolvers = ['9.9.9.11:53', '8.8.8.8:53']`

2. "although you disabled system DNS .."

 [resolv.conf](https://github.com/llightcb/sight/blob/main/install.sh#L204)

(margin note: [+egress](https://github.com/llightcb/sight/blob/261aa143ef376c839adb0b0d8012f8a20acb9c2d/home/.config/fish/functions/fwn.fish#L68), because chromium, for example, has _async dns_  (built-in dns resolver) enabled _by default_ - _as far as i remember_ (my chr. dns-flags are explicitly set since 2 years). firefox -  probably - behaves similarly. and so forth. 

3. it happens every 24 hours. 
**q**: but **why**? [because](https://github.com/DNSCrypt/dnscrypt-proxy/discussions/2201#discussioncomment-3688991)?

4. endnote: i'm - again - tempted to draw a line over doh (you may remember) - if there wasn't only quad9 as a _dnscrypt_ alternative to cloudflare security - and my trust in quad9 is down to zero (since a "strange" happening) . 2nd side note: doh3 is not progressing either (resp. finally _+ alt-svc header_ to the responses (@ cloudflare @ quad9 : do your job).

ps: additional info. i just looked at my logs from july. sometimes this line also appeared (august: not anymore): 

`unable to resolve [ipv6.download.dnscrypt.info] - make sure that the system resolver works, or that bootstrap_resolvers has been set to resolvers that can be reached`

of course, i disabled ipv6 on my system.  anyway:

`bootstrap_resolvers = ['9.9.9.11:53', '8.8.8.8:53']`

 today - right now - after 24h:

`bootstrap resolvers didn't respond - trying with the system resolver as a last resort`

last line of the whole "thing".

**Reply by lifenjoiner** [reply]
Reviewed the [v2.1.4 source code](https://github.com/DNSCrypt/dnscrypt-proxy/blob/9f3ef735f283a90d97b041a4d1d6bf8f866fe6ef/dnscrypt-proxy/xtransport.go#L365-L404) (there has been changes), and found I made a mistake. So, let's make it clear what happened, afresh:

The log items should be grouped like this:
```
Jul 26 14:00:03 System DNS configuration not usable yet, exceptionally resolving [download.dnscrypt.net] using bootstrap resolvers over tcp
Jul 26 14:00:07 Bootstrap resolvers didn't respond - Trying with the system resolver as a last resort
...
```
line 1:
1. `System DNS configuration not usable yet`, means you set `ignore_system_dns = true`, so it did `using bootstrap resolvers over tcp`;
2. `tcp` meaning you must set [force tcp](https://github.com/DNSCrypt/dnscrypt-proxy/blob/9f3ef735f283a90d97b041a4d1d6bf8f866fe6ef/dnscrypt-proxy/example-dnscrypt-proxy.toml#L97), but not mentioned. It tried TCP first, failed, showed the log, and then UDP, still failed.

line 2:
`Bootstrap resolvers didn't respond`, timeout. At last (although you disabled system DNS, but there is no other ways left), it tried `with the system resolver as a last resort`, and succeeded.

Downloading source file starts.

<br>

Extra info:
download.dnscrypt.**net** is not in the latest URLs.
1. You are not use the latest version. There are some important bug fixes. Maybe the bug causing this has been fixed, which we wouldn't dive in.
2. Maybe it is why there are more than 2 groups of failures. No source file on this site anymore, one more retry on other URL.

`resolv.conf` implies you're more likely using an old version.

<br>

Why `Bootstrap resolvers didn't respond`?
That may be hard to say. You should make sure them work well and stable in the normally way, for example with `dig`, and then we may narrow the issue to `dnscrypt-proxy`. But, considering you are using an old version, I suggest you upgrade it first.

<br>

> it happens every 24 hours.
> q: but why? https://github.com/DNSCrypt/dnscrypt-proxy/discussions/2201#discussioncomment-3688991?

`dnscrypt-proxy` update the source files every 24h, forcibly.

<br>

PS:
If there is any issue, you'd better provide information like the issue template requests: OS version, software version, config, logs and etc. which may give plenty of information and help reproduce your issue. That will help not only us who want to help you, but also yourself in fact.

**Reply by llightcb** [reply]
thanks again for the quick reply + for looking @ code and co. .

why does the resolv.conf indicate that i have an old dnscrypt-proxy version? i don't see a causal connection. i will do an `apk fix dnscrypt-proxy` before. the version is up to date, but it could be that i have an outdated .toml (the one from previous versions, not: `dnscrypt-proxy.toml.apk-new`). a diff between these 2 files will also be interesting.

**Reply by lifenjoiner** [reply]
There is too few information you give, as I said at the beginning.
I can't help you anymore but just guess.

**Reply by llightcb** [reply]
i will use the new .toml first. but a diff between these 2 configs i did 5 min ago showed me that this will be useless. no difference that would matter. again: version is up to date. et cetera. see you (maybe) in _issues_, in time, when i am not _that_ tired anymore.

regarding: [download.dnscrypt.net](https://github.com/DNSCrypt/dnscrypt-proxy/blob/1792c06bc709c67e983075941841a0befb2ee5a4/ChangeLog#L121) ? that's the actual url of the actual version. i hope everything is ok with this download mirror.

**Reply by lifenjoiner** [reply]
Just recall all my guesses due to no guarantees but goes too far away.

"like the issue template requests" is not saying you must log an issue. Wherever you want someone to dive into something, you must explain your problem clear first. Normally, redundant information is requested to make sure enough, and, the fact may not be what you think.

Say:
Tom used to eat an apple every day. But he left it untouched last week. Why?

It's a pity that I don't have this problem nor can reproduce it. I don't have the answer. Let's see what others say.
